### PR TITLE
Bug 1891314: Return helm charts based on the installed kubernetes version

### PR DESCRIPTION
**Pull request by akashshinde**
- This PR enables support to filter charts based on the kubernetes version. 
- `onlyCompatible` flag is provided to `/api/helm/charts/index.yaml` API which drops the unsupported charts from the response.

Ref: https://issues.redhat.com/browse/HELM-12

### CODE REVIEW

File: pkg/helm/chartproxy/proxy.go [hunk 1]
```diff
@@ -44,13 +48,14 @@ func coreClientProvider(p *proxy) error {
 
 var defaultOptions = []ProxyOption{dynamicKubeClientProvider, coreClientProvider}
 
-func New(k8sConfig RestConfigProvider, opts ...ProxyOption) (Proxy, error) {
+func New(k8sConfig RestConfigProvider, helmConfig *action.Configuration, opts ...ProxyOption) (Proxy, error) {
```
[the remark this answers is not in the document]
Comment (akashshinde) [reply pkg/helm/chartproxy/proxy.go]: Fixed

File: pkg/helm/chartproxy/proxy.go [hunk 2]
```diff
@@ -44,13 +48,14 @@ func coreClientProvider(p *proxy) error {
...
 	config, err := k8sConfig()
 	if err != nil {
 		return nil, err
 	}
 	p := &proxy{
-		config: config,
+		config:     config,
+		helmConfig: helmConfig,
```
[the remark this answers is not in the document]
Comment (akashshinde) [reply pkg/helm/chartproxy/proxy.go]: Fixed, please take a look at updated changes

File: pkg/helm/chartproxy/proxy.go [hunk 3]
```diff
@@ -71,15 +76,54 @@ func (p *proxy) IndexFile() (*repo.IndexFile, error) {
 	}
 
 	indexFile := repo.NewIndexFile()
+	kubeVersion, err := p.GetKubeVersions()
```
[the remark this answers is not in the document]
Comment (akashshinde) [reply pkg/helm/chartproxy/proxy.go]: Done

File: pkg/helm/chartproxy/proxy_test.go [hunk 4]
```diff
@@ -16,16 +17,29 @@ import (
...
 func TestProxy_IndexFile(t *testing.T) {
 	tests := []struct {
-		name       string
-		indexFiles []string
-		mergedFile string
-		helmCRS    []*unstructured.Unstructured
+		name           string
+		indexFiles     []string
+		mergedFile     string
+		kubeVersion    *chartutil.KubeVersion
```
Comment (pedjak) [inline pkg/helm/chartproxy/proxy_test.go:23]: could we use string here, and tests values according to the doc syntax: https://helm.sh/docs/topics/charts/#the-kubeversion-field ?

File: pkg/helm/chartproxy/proxy_test.go [hunk 5]
```diff
@@ -16,16 +17,29 @@ import (
...
+		helmCRS        []*unstructured.Unstructured
+		onlyCompatible bool
 	}{
 		{
 			name:       "returned index file for configured helm repo",
 			indexFiles: []string{"testdata/sampleRepoIndex.yaml"},
 			mergedFile: "testdata/mergedSampleRepoIndex2.yaml",
 		},
+		{
+			name:       "returned index file for configured helm repo for provided kubernetes cluster filter",
```
Comment (pedjak) [inline pkg/helm/chartproxy/proxy_test.go:33]: how about: retuned index file for configured helm repo contains only charts compatible for given cluster?

File: pkg/helm/chartproxy/proxy_test.go [hunk 6]
```diff
@@ -44,7 +58,7 @@ func TestProxy_IndexFile(t *testing.T) {
 		{
 			name:       "returned merged index file for all accessible helm repos",
 			indexFiles: []string{"testdata/azureRepoIndex.yaml"},
-			mergedFile: "testdata/azureRepoIndex.yaml",
+			mergedFile: "testdata/mergedAzureRepoIndex.yaml",
```
Comment (pedjak) [inline pkg/helm/chartproxy/proxy_test.go:61]: why this change here? we do not apply any filter here.

File: pkg/helm/chartproxy/proxy_test.go [hunk 7]
```diff
@@ -103,7 +121,7 @@ func TestProxy_IndexFile(t *testing.T) {
 					t.Error(err)
 				}
 				if !reflect.DeepEqual(indexFile.Entries, expectedIndexFile.Entries) {
-					t.Errorf("Expected index content \n%v but got \n%v", expectedIndexFile, indexFile)
+					t.Errorf("Expected index content \n%+v but got \n%+v", expectedIndexFile, indexFile)
```
Comment (pedjak) [inline pkg/helm/chartproxy/proxy_test.go:124]: why this change?

File: pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml [hunk 8]
```diff
@@ -0,0 +1,164 @@
+apiVersion: v1
```
Comment (pedjak) [inline pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml:1]: do we really need this file?

File: pkg/helm/chartproxy/proxy_test.go (hunk 5 above)
Comment (pedjak) [reply pkg/helm/chartproxy/proxy_test.go:33]: we should also have a test where we none of charts are compatible with given cluster, i.e. the returned index should be empty.

File: pkg/helm/handlers/handler_test.go [hunk 9]
```diff
@@ -681,25 +686,30 @@ func TestHelmHandlers_Index(t *testing.T) {
 		proxyNewError    error
 		indexFileError   error
 		expectedResponse string
+		urlQuery         string
+		onlyCompatible bool
 	}{
 		{
-			name:           "valid repo index file should return correct response",
-			httpStatusCode: http.StatusOK,
+			name: "valid repo index file should return correct response",
```
Comment (pedjak) [inline pkg/helm/handlers/handler_test.go:693]: why this change? we should have another test with filtering set to on.. and this test should remain untouched.

File: pkg/helm/handlers/handler_test.go [hunk 10]
```diff
@@ -719,18 +729,23 @@ func TestHelmHandlers_Index(t *testing.T) {
...
-			request = httptest.NewRequest(http.MethodGet, "/api/helm/charts/index.yaml", strings.NewReader(""))
+			request = httptest.NewRequest(http.MethodGet, "/api/helm/charts/index.yaml"+tt.urlQuery, strings.NewReader(""))
 			response := httptest.NewRecorder()
 
 			user := &auth.User{
 				Token: "foo",
 			}
 			handler := helmHandlers{
-				newProxy: func(token string) (proxy chartproxy.Proxy, err error) {
+				newProxy: func(token string, opts ...chartproxy.ProxyOption) (proxy chartproxy.Proxy, err error) {
```
Comment (pedjak) [inline pkg/helm/handlers/handler_test.go:739]: is `opts` used anywhere?

**Review by pedjak — CHANGES_REQUESTED** [review]

File: pkg/helm/handlers/handlers.go [hunk 11]
```diff
@@ -32,18 +34,27 @@ func New(apiUrl string, transport http.RoundTripper) *helmHandlers {
 		rollbackRelease:         actions.RollbackRelease,
 		getReleaseHistory:       actions.GetReleaseHistory,
 	}
-	h.newProxy = func(bearerToken string) (getter chartproxy.Proxy, err error) {
+
+	version, err := cluster.GetKubeVersion(h.restConfig(""))
```
[the remark this answers is not in the document]
Comment (akashshinde) [reply pkg/helm/handlers/handlers.go:38]: Done

File: pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml (hunk 8 above)
Comment (akashshinde) [reply pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml:1]: adding this correct index file fixes the failing unit test

File: pkg/helm/handlers/handler_test.go (hunk 10 above)
Comment (akashshinde) [reply pkg/helm/handlers/handler_test.go]: removed

File: pkg/helm/chartproxy/proxy_test.go (hunk 6 above)
Comment (akashshinde) [reply pkg/helm/chartproxy/proxy_test.go:63]: Need to add this file to fix failing unit test

File: pkg/helm/chartproxy/proxy_test.go [hunk 12]
```diff
@@ -1,31 +1,47 @@
...
 		},
+		{
+			name:           "returned index file for configured helm repo contains only charts compatible for given cluster",
+			indexFiles:     []string{"testdata/sampleRepoIndex3.yaml"},
+			mergedFile:     "testdata/sampleRepoIndex3Filtered.yaml",
+			kubeVersion:    "v1.16.0",
+			onlyCompatible: true,
+		},
+		{
+			name:           "returned empty index file for configured helm repo contains only charts compatible for given cluster",
```
Comment (pedjak) [inline pkg/helm/chartproxy/proxy_test.go:53]: let's name it as `return empty index file if not charts are compatible with given cluster`

File: pkg/helm/chartproxy/proxy_test.go (hunk 7 above)
Comment (pedjak) [reply pkg/helm/chartproxy/proxy_test.go:137]: @akashshinde ?

File: pkg/helm/cluster/kube_version.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
+package cluster
```
Comment (pedjak) [inline pkg/helm/cluster/kube_version.go:1]: do we really need a new package for just single function?

File: pkg/helm/handlers/handler_test.go [hunk 14]
```diff
@@ -700,6 +706,29 @@ func TestHelmHandlers_Index(t *testing.T) {
 					},
 				},
 			},
+			httpStatusCode: http.StatusOK,
+		},
+		{
+			name: "valid repo index file should return correct response and also validate onlyCompatible flag",
```
Comment (pedjak) [inline pkg/helm/handlers/handler_test.go:712]: "valid repo index file should contain only entries compatible with the given cluster"

File: pkg/helm/handlers/handler_test.go [hunk 15]
```diff
@@ -719,7 +748,7 @@ func TestHelmHandlers_Index(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			var request *http.Request
 
-			request = httptest.NewRequest(http.MethodGet, "/api/helm/charts/index.yaml", strings.NewReader(""))
+			request = httptest.NewRequest(http.MethodGet, "/api/helm/charts/index.yaml"+tt.urlQuery, strings.NewReader(""))
```
Comment (pedjak) [inline pkg/helm/handlers/handler_test.go:751]: we should also have a test where `onlyCompatible` is send from client and the value is `false`.

File: pkg/helm/handlers/handlers.go [hunk 16]
```diff
@@ -255,7 +266,11 @@ func (h *helmHandlers) HandleGetReleaseHistory(user *auth.User, w http.ResponseW
 }
 
 func (h *helmHandlers) HandleIndexFile(user *auth.User, w http.ResponseWriter, r *http.Request) {
-	proxy, err := h.newProxy(user.Token)
+	var proxy chartproxy.Proxy
+	var err error
```
[the remark this answers is not in the document]
Comment (pedjak) [reply pkg/helm/handlers/handlers.go:262]: @akashshinde ^^

File: pkg/server/server.go [hunk 17]
```diff
@@ -18,18 +21,17 @@ import (
 	"github.com/coreos/dex/api"
 	"github.com/coreos/pkg/health"
 
+	graphql "github.com/graph-gophers/graphql-go"
+	"github.com/graph-gophers/graphql-go/relay"
+	"github.com/rawagner/graphql-transport-ws/graphqlws"
+
 	"github.com/openshift/console/pkg/auth"
+	"github.com/openshift/console/pkg/graphql/resolver"
```
Comment (pedjak) [inline pkg/server/server.go:29]: please revert import sorting

File: pkg/server/server.go [hunk 18]
```diff
@@ -389,8 +391,23 @@ func (s *Server) HTTPHandler() http.Handler {
...
+	})
+	if err != nil {
+		klog.Warningf("Failed to get kube discovery client %s", err.Error())
+	}
+
+	kubeVersion, err := client.ServerVersion()
+	if err != nil {
+		klog.Warningf("Failed to get server version %s", err.Error())
+	}
+
```
Comment (pedjak) [inline pkg/server/server.go:407]: we should wrap this up into a function and move it under `helm` package

**Review by pedjak — CHANGES_REQUESTED** [review]

**Comment by pedjak** [comment]
`pkg/helm/chartproxy/testdata/emptyRepoIndex.yaml` is empty and not used anywhere..

**Comment by rohitkrai03** [comment]
@akashshinde @pedjak I would prefer if we make it default that all chart are filtered based on compatibility as this is the direction that we are looking at for Helm catalog by default. It also removes the need to change the UI for this. We can have an option to send all charts unfiltered that UI can use when required but there is no need for fetching unfiltered charts right now so it would be great if we can consider that.

**Comment by pedjak** [comment]
@rohitkrai03 thanks for the feedback.. we can revert the default behaviour...

File: pkg/helm/handlers/handler_test.go (hunk 9 above)
Comment (akashshinde) [reply pkg/helm/handlers/handler_test.go:693]: Added new test.

File: pkg/helm/chartproxy/proxy_test.go (hunk 7 above)
Comment (akashshinde) [reply pkg/helm/chartproxy/proxy_test.go:137]: reverted

File: pkg/helm/chartproxy/proxy_test.go (hunk 12 above)
Comment (akashshinde) [reply pkg/helm/chartproxy/proxy_test.go:53]: Done

File: pkg/helm/cluster/kube_version.go (hunk 13 above)
Comment (akashshinde) [reply pkg/helm/cluster/kube_version.go:1]: Operation of getting kube version is not fitting anywhere in the packages we have in `helm` package, therefore created this package to get cluster related information.

Comment (akashshinde) [reply pkg/helm/cluster/kube_version.go:1]: I am okay with having this any other package in helm if that fits well.

File: pkg/helm/handlers/handler_test.go (hunk 14 above)
Comment (akashshinde) [reply pkg/helm/handlers/handler_test.go:712]: Done

File: pkg/helm/handlers/handler_test.go (hunk 15 above)
Comment (akashshinde) [reply pkg/helm/handlers/handler_test.go:751]: added

File: pkg/server/server.go (hunk 18 above)
Comment (akashshinde) [reply pkg/server/server.go:407]: Done, added it under `actions` packages, let me know if that doesn't fit well.

File: pkg/helm/chartproxy/proxy_test.go [hunk 19]
```diff
@@ -44,7 +74,7 @@ func TestProxy_IndexFile(t *testing.T) {
 		{
 			name:       "returned merged index file for all accessible helm repos",
 			indexFiles: []string{"testdata/azureRepoIndex.yaml"},
-			mergedFile: "testdata/azureRepoIndex.yaml",
+			mergedFile: "testdata/mergedAzureRepoIndex.yaml",
```
Comment (pedjak) [inline pkg/helm/chartproxy/proxy_test.go:77]: revert this change now, give that we have separate PR #7111

File: pkg/helm/chartproxy/proxy_test.go (hunk 7 above)
Comment (pedjak) [reply pkg/helm/chartproxy/proxy_test.go:137]: still see it in the code.

**Review by pedjak — CHANGES_REQUESTED** [review]
please look at the existing comments, there are still some unaddressed.

File: pkg/helm/chartproxy/proxy_test.go (hunk 19 above)
Comment (akashshinde) [reply pkg/helm/chartproxy/proxy_test.go]: Done

File: pkg/helm/actions/kube_version.go [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+
+	kubeVersion, err := client.ServerVersion()
+	if err != nil {
+		return "", err
+	}
+
+	if kubeVersion != nil {
+		return kubeVersion.String(), errors.New("failed to get kubernetes version")
+	}
+	return "", nil
```
Comment (pedjak) [inline pkg/helm/actions/kube_version.go:28]: should we return error here instead?

File: pkg/helm/actions/kube_version.go [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+		return "", err
+	}
+
+	kubeVersion, err := client.ServerVersion()
+	if err != nil {
+		return "", err
+	}
+
+	if kubeVersion != nil {
+		return kubeVersion.String(), errors.New("failed to get kubernetes version")
```
Comment (pedjak) [inline pkg/helm/actions/kube_version.go:26]: hm, if kubeversion is not nill, then we have it, we should not return the error as well?

File: pkg/helm/actions/kube_version.go [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"errors"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"net/http"
+)
+
+func GetKubeVersion(apiEndpoint string, transport http.RoundTripper) (string, error) {
+	config := &rest.Config{
```
Comment (pedjak) [inline pkg/helm/actions/kube_version.go:11]: why not passing to the function `rest.Config` instead?

File: pkg/helm/cluster/kube_version.go [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"errors"
+	"helm.sh/helm/v3/pkg/chartutil"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+)
+
+func GetKubeVersion(config *rest.Config) (*chartutil.KubeVersion, error) {
```
Comment (pedjak) [inline pkg/helm/cluster/kube_version.go:10]: should not we remove this function here, since you have introduced one in `actions` package?

File: pkg/helm/handlers/handlers.go [hunk 24]
```diff
@@ -255,6 +256,9 @@ func (h *helmHandlers) HandleGetReleaseHistory(user *auth.User, w http.ResponseW
 }
 
 func (h *helmHandlers) HandleIndexFile(user *auth.User, w http.ResponseWriter, r *http.Request) {
+	// Setting this by default to true, this always serves helm index file with compatible chart lists.
+	onlyCompatible := true
```
Comment (pedjak) [inline pkg/helm/handlers/handlers.go:260]: that should be default behaviour, but we should have a possibility to set this to `false`. We should have tests that verify that as well.

File: pkg/server/server.go [hunk 25]
```diff
@@ -389,8 +392,15 @@ func (s *Server) HTTPHandler() http.Handler {
 	handle("/api/console/knative-channels", authHandler(s.handleKnativeChannelCRDs))
 	handle("/api/console/version", authHandler(s.versionHandler))
 
+	// Get running kubernetes version and inject into helm handlers.
+	kubeVersion, err := actions.GetKubeVersion(s.KubeAPIServerURL, s.K8sClient.Transport)
+	if err != nil {
+		klog.Warningf("Failed to get server version %s", err.Error())
```
Comment (pedjak) [inline pkg/server/server.go:398]: how about "Failed to get cluster k8s version"?

File: pkg/server/server.go [hunk 26]
```diff
@@ -389,8 +392,15 @@ func (s *Server) HTTPHandler() http.Handler {
 	handle("/api/console/knative-channels", authHandler(s.handleKnativeChannelCRDs))
 	handle("/api/console/version", authHandler(s.versionHandler))
 
+	// Get running kubernetes version and inject into helm handlers.
+	kubeVersion, err := actions.GetKubeVersion(s.KubeAPIServerURL, s.K8sClient.Transport)
```
Comment (pedjak) [inline pkg/server/server.go:396]: please use `s.K8sProxyConfig.Endpoint.String()` instead.

File: pkg/server/server.go [hunk 27]
```diff
@@ -389,8 +392,15 @@ func (s *Server) HTTPHandler() http.Handler {
...
+	if err != nil {
+		klog.Warningf("Failed to get server version %s", err.Error())
+	}
+
+	helmHandlers := helmhandlerspkg.New(s.KubeAPIServerURL, s.K8sClient.Transport, kubeVersion)
```
Comment (pedjak) [inline pkg/server/server.go:401]: please use s.K8sProxyConfig.Endpoint.String() instead.

File: pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml (hunk 8 above)
Comment (pedjak) [inline pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml:1]: we do not need this file anymore in this PR?

File: docs/helm/endpoints_api.md [hunk 28]
```diff
@@ -334,6 +334,9 @@ _Returns repository index file containing all entries from all configured reposi
 * **Method:**
 
   `GET`
+  
+* **Supported URL Query Parameter:**
+    *  `onlyCompatible` - `true`/`false` Setting true would return helm charts which are supported in the provided cluster.
```
Comment (pedjak) [inline docs/helm/endpoints_api.md:339]: we should state what is the default value if omitted.

**Review by pedjak — CHANGES_REQUESTED** [review]

**Comment by pedjak** [comment]
/retitle Bug 1891314: Return helm charts based on the installed kubernetes version

File: pkg/server/server.go (hunk 25 above)
Comment (akashshinde) [reply pkg/server/server.go:398]: done

File: pkg/helm/handlers/handlers.go (hunk 24 above)
Comment (akashshinde) [reply pkg/helm/handlers/handlers.go:260]: done

File: pkg/helm/cluster/kube_version.go (hunk 23 above)
Comment (akashshinde) [reply pkg/helm/cluster/kube_version.go:10]: done

File: pkg/server/server.go (hunk 26 above)
Comment (akashshinde) [reply pkg/server/server.go:396]: done

File: pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml (hunk 8 above)
Comment (akashshinde) [reply pkg/helm/chartproxy/testdata/mergedAzureRepoIndex.yaml:1]: removed

File: pkg/server/server.go (hunk 27 above)
Comment (akashshinde) [reply pkg/server/server.go:401]: done

File: pkg/helm/actions/kube_version.go (hunk 22 above)
Comment (akashshinde) [reply pkg/helm/actions/kube_version.go:11]: done

File: pkg/helm/actions/kube_version.go (hunk 21 above)
Comment (akashshinde) [reply pkg/helm/actions/kube_version.go:26]: no need to return error here, thanks
done

File: pkg/helm/actions/kube_version.go (hunk 20 above)
Comment (akashshinde) [reply pkg/helm/actions/kube_version.go:28]: done

File: docs/helm/endpoints_api.md (hunk 28 above)
Comment (akashshinde) [reply docs/helm/endpoints_api.md:339]: done

File: pkg/server/server.go [hunk 29]
```diff
@@ -389,8 +393,19 @@ func (s *Server) HTTPHandler() http.Handler {
...
+	config := &rest.Config{
+		APIPath:   s.K8sProxyConfig.Endpoint.String(),
+		Transport: s.K8sClient.Transport,
+	}
+	kubeVersion, err := actions.GetKubeVersion(config)
+	if err != nil {
+		klog.Warningf("Failed to get cluster k8s version %s", err.Error())
+	}
+
+	helmHandlers := helmhandlerspkg.New(s.KubeAPIServerURL, s.K8sClient.Transport, kubeVersion)
```
Comment (pedjak) [inline pkg/server/server.go:406]: This line should also use `s.K8sProxyConfig.Endpoint.String()` not `s.KubeAPIServerURL`. Check the removed line.

Comment (akashshinde) [reply pkg/server/server.go:406]: done

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@akashshinde @pedjak I don't see any test cases that tests for pre-release version of Kubernetes. The main reason for moving this logic from frontend to backend was so that the pre-release versions work correctly while filtering.

**Comment by pedjak** [comment]
> I don't see any test cases that tests for pre-release version of Kubernetes. The main reason for moving this logic from frontend to backend was so that the pre-release versions work correctly while filtering.

@rohitkrai03 please suggest/craft some index.yaml and share with us some the expectations how it should work. Thanks!

**Review by pedjak — APPROVED** [review]
lgtm but let add additional tests asked by @rohitkrai03

**Comment by akashshinde** [comment]
@rohitkrai03 added few tests to check pre-release kube version compatibility. https://github.com/openshift/console/pull/7012/commits/fe7d52ae1c510aa8337ff02670f194875d6a4d24

File: pkg/helm/chartproxy/testdata/RepoIndexPreRelease.yaml [hunk 30]
```diff
@@ -0,0 +1,78 @@
...
+      keywords:
+        - amd64
+        - deploy
+        - Catalog
+        - Commercial
+        - Limited
+        - Other
+        - RHOCP
+      kubeVersion: '>=1.20.0-beta.2'
```
Comment (rohitkrai03) [inline pkg/helm/chartproxy/testdata/RepoIndexPreRelease.yaml:70]: Usually a helm chart that support pre-release versions specifies kubeVersion range with a `-0` at the end. Below is a helm chart metadata from our catalog that support pre-release version - 

```yaml
- apiVersion: v2
    appVersion: 10.0.0
    created: "2020-11-03T14:58:13.7979+05:30"
    dependencies:
    - alias: sch
      name: ibm-sch
      repository: '@sch'
      version: 1.2.15
    description: |-
      IBM Order management Software Enterprise Edition v10 provides cross-channel order orchestration capabilities to enable intelligent brokering of orders across many disparate systems. It also provides a global view of inventory across the supply chain and enables business users to make changes to order process.
      Documentation
      For additional details regarding install parameters check http://ibm.biz/oms-helm-readme.
      License
      By installing this product you accept the license terms http://ibm.biz/oms-license and http://ibm.biz/oms-apps-license.
    digest: 47f7bae2c23ceec955c497ddd3f1d2adeea08238363a849c3ffc0ebefb9125ff
    home: http://ibm.biz/oms-home
    icon: https://raw.githubusercontent.com/IBM/charts/master/logo/ibm-oms-logo.png
    keywords:
    - oms
    - sterling
    - yantra
    - order
    - fulfillment
    - om
    - amd64
    - ppc64le
    - framework
    - Commercial
    - RHOCP
    - Other
    kubeVersion: '>=1.16.0-0'
    maintainers:
    - name: IBM
    name: ibm-oms-ent-prod
    type: application
    urls:
    - https://redhat-developer.github.com/redhat-helm-charts/charts/ibm-oms-ent-prod-5.1.0.tgz
    version: 5.1.0
```

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: pkg/helm/chartproxy/testdata/RepoIndexPreRelease.yaml (hunk 30 above)
Comment (akashshinde) [reply pkg/helm/chartproxy/testdata/RepoIndexPreRelease.yaml:70]: Added https://github.com/openshift/console/pull/7012/commits/81e25d75997565ad4514d6dc224def24290044d9
cc: @rohitkrai03

**Comment by rohitkrai03** [comment]
@akashshinde I created a `4.6.0-0.nightly-2020-09-03-191144` cluster to test this PR. It's the same cluster that the [bug](https://bugzilla.redhat.com/show_bug.cgi?id=1876741) was reported. 

```
❯ oc version
Client Version: 4.5.5
Server Version: 4.6.0-0.nightly-2020-09-03-191144
Kubernetes Version: v4.6.0-202009031349.p0-dirty

```

Added below `HelmChartRepository` CR - 
```yaml
apiVersion: helm.openshift.io/v1beta1
kind: HelmChartRepository
metadata:
  name: ibm-sample-repo
spec:
  name: ibm-sample-repo
  connectionConfig:
    url: https://raw.githubusercontent.com/IBM/charts/master/repo/ibm-helm
```
The repo has in total 38 charts in it's index.yaml. Out of 38 only 5 specifies `kubeVersion` with `-0` at the end which means only 5 support pre release kubeVersion.

Right now in the console i am getting 12 Helm charts from the backend out of which only 3 specifies `kubeVersion` with `-0`. Not sure how the filtering it working at the moment but i think there might be some issue here with the filtering.

**Comment by akashshinde** [comment]
> On a pre-release kubernetes cluster I would expect only those charts to show up which have -0 specified.

If kubernetes pre-release cluster version is `v1.19.1-beta.3` then following charts can be listed.

- kube >= v1.19.0 
- kube >= v1.19.1
- kube >= v1.19.1-0 
- kube == v1.19.1-beta3

**Comment by rohitkrai03** [comment]
After discussions with @akashshinde it seems the current filtering logic should work for all of our use cases. The above shared cluster has something wrong because it shows `Kubernetes Version: v4.6.0-202009031349.p0-dirty` which is not possible.

**Review by rohitkrai03 — APPROVED** [review]
/lgtm

**Comment by akashshinde** [comment]
@spadgett Can you take a look at this ? There are few changes in `server.go` which requires your attention.

**Comment by akashshinde** [comment]
@christianvogt can you please take a look at this ?

File: pkg/server/server.go [hunk 31]
```diff
@@ -389,8 +393,19 @@ func (s *Server) HTTPHandler() http.Handler {
...
 	handle("/api/console/version", authHandler(s.versionHandler))
 
+	// Get running kubernetes version and inject into helm handlers.
+	config := &rest.Config{
+		APIPath:   s.K8sProxyConfig.Endpoint.String(),
+		Transport: s.K8sClient.Transport,
+	}
+	kubeVersion, err := actions.GetKubeVersion(config)
+	if err != nil {
+		klog.Warningf("Failed to get cluster k8s version %s", err.Error())
```
Comment (spadgett) [inline pkg/server/server.go:403]: We'll want to make sure this is logged by default. @jhadvig Do you know offhand?

File: pkg/server/server.go [hunk 32]
```diff
@@ -389,8 +393,19 @@ func (s *Server) HTTPHandler() http.Handler {
 	handle("/api/console/knative-channels", authHandler(s.handleKnativeChannelCRDs))
 	handle("/api/console/version", authHandler(s.versionHandler))
 
+	// Get running kubernetes version and inject into helm handlers.
+	config := &rest.Config{
+		APIPath:   s.K8sProxyConfig.Endpoint.String(),
+		Transport: s.K8sClient.Transport,
+	}
+	kubeVersion, err := actions.GetKubeVersion(config)
```
Comment (spadgett) [inline pkg/server/server.go:401]: I guess it's OK to check only once on startup since the API server is upgraded before the other cluster operators. There might be a small window during upgrades where we show charts for the previous version after the API servers are upgraded.

Comment (spadgett) [reply pkg/server/server.go:401]: Note that I believe the kube version is available as an environment variable in the pod, so we could conceivably avoid needing to make this request (although not for off-cluster mode).

File: pkg/server/server.go (hunk 31 above)
Comment (spadgett) [reply pkg/server/server.go:403]: We should fall back to the env var at least on failure.

File: pkg/server/server.go (hunk 32 above)
Comment (akashshinde) [reply pkg/server/server.go:413]: @spadgett  Getting kube version from env var `KUBE_GIT_VERSION` in case api server failed to report. Done https://github.com/openshift/console/pull/7012/commits/739ad74f846e52fced4420d57038ce65ef98f449

**Review by spadgett — APPROVED** [review]
/lgtm
